x-pack/plugin/apm-data/src/main/java/org/elasticsearch/xpack/apmdata/APMPlugin.java

```diff
@@ -10,36 +10,62 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.apache.lucene.util.SetOnce;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.plugins.ActionPlugin;
 import org.elasticsearch.plugins.Plugin;
+import org.elasticsearch.xpack.core.XPackSettings;
 
 import java.util.Collection;
+import java.util.Collections;
 import java.util.List;
 
 public class APMPlugin extends Plugin implements ActionPlugin {
     private static final Logger logger = LogManager.getLogger(APMPlugin.class);
 
-    private final SetOnce<APMIndexTemplateRegistry> registry = new SetOnce<>();
+    final SetOnce<APMIndexTemplateRegistry> registry = new SetOnce<>();
+
+    private final boolean enabled;
+
+    // APM_DATA_REGISTRY_ENABLED controls enabling the index template registry.
+    //
+    // This setting will be ignored if the plugin is disabled.
+    static final Setting<Boolean> APM_DATA_REGISTRY_ENABLED = Setting.boolSetting(
+        "xpack.apm_data.registry.enabled",
+        true,
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+
+    public APMPlugin(Settings settings) {
+        this.enabled = XPackSettings.APM_DATA_ENABLED.get(settings);
+    }
 
     @Override
     public Collection<?> createComponents(PluginServices services) {
+        logger.info("APM ingest plugin is {}", enabled ? "enabled" : "disabled");
+        Settings settings = services.environment().settings();
+        ClusterService clusterService = services.clusterService();
         registry.set(
-            new APMIndexTemplateRegistry(
-                services.environment().settings(),
-                services.clusterService(),
-                services.threadPool(),
-                services.client(),
-                services.xContentRegistry()
-            )
+            new APMIndexTemplateRegistry(settings, clusterService, services.threadPool(), services.client(), services.xContentRegistry())
         );
-        APMIndexTemplateRegistry registryInstance = registry.get();
-        logger.info("APM ingest plugin is {}", registryInstance.isEnabled() ? "enabled" : "disabled");
-        registryInstance.initialize();
-        return List.of(registryInstance);
+        if (enabled) {
+            APMIndexTemplateRegistry registryInstance = registry.get();
+            registryInstance.setEnabled(APM_DATA_REGISTRY_ENABLED.get(settings));
+            clusterService.getClusterSettings().addSettingsUpdateConsumer(APM_DATA_REGISTRY_ENABLED, registryInstance::setEnabled);
```

Comment:
We tend to encapsulate this in the code of the class using the `enabled` flag, in this case the `APMIndexTemplateRegistry`. We usually read the value of the setting and we set it in the constructor and in a method like `initialize()` we register the listener. This way it's transparent in the `APMIndexTemplateRegistry` how the enabled flag is changing.

Reply:
@gmarouli is it worth me changing now? I can go back and update it.

The only way the `enabled` flag is updated is through the `setEnabled` method, and I'm not sure why `APMIndexTemplateRegistry` would care what caused that method to be called -- but if it's conventional and this difference will be confusing, I can go back and update it.

Reply:
It's more about encapsulation, the logic of how the flag for this class is changing is contained in the class itself. It's ok for now, if we come back to this we can address it.